### PR TITLE
Feature/fip 0027 market label

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
FIP-0027 - Label encoding for market actors v8 

### Alternate Designs *[optional]*
Actors v8 to be implemented.

### CODE REVIEW

File: core/vm/actor/builtin/types/market/deal.hpp [hunk 1]
```diff
@@ -119,7 +59,7 @@ namespace fc::vm::actor::builtin::types::market {
   CBOR_TUPLE(DealState, sector_start_epoch, last_updated_epoch, slash_epoch)
 
   struct ClientDealProposal {
-    DealProposal proposal;
+    Universal<DealProposal> proposal;
```
Comment (Alexey-N-Chernyshov) [inline core/vm/actor/builtin/types/market/deal.hpp:62]: `proposal` is ptr now and default copy assigment of `ClientDealProposal` is a shallow copy.
Please, take a look.

File: core/api/full_node/make.cpp [hunk 2]
```diff
@@ -42,15 +42,11 @@
...
-  }
+#define MOVE(x) \
+  x { std::move(x) }
 
-#define FWD(x)                   \
-  x {                            \
-    std::forward<decltype(x)>(x) \
-  }
+#define FWD(x) \
+  x { std::forward<decltype(x)>(x) }
```
Comment (turuslan) [inline core/api/full_node/make.cpp:49]: Is it clang-format 11?  
My clang-format 11.0.0 formats as old (red), so it will revert this change.

File: core/vm/actor/builtin/types/universal/universal.hpp [hunk 3]
```diff
@@ -25,9 +25,30 @@ namespace fc::vm::actor::builtin::types {
...
+      object = other.object;
+      return *this;
+    }
+
+    Universal<T> &operator=(Universal<T> &&other) noexcept {
+      actor_version = other.actor_version;
+      object = std::move(other.object);
+      return *this;
+    }
+
```
Comment (turuslan) [inline core/vm/actor/builtin/types/universal/universal.hpp:51]: Are implicit constructor/operator not working?

File: core/vm/actor/builtin/types/market/deal.hpp (hunk 1 above)
Comment (wer1st) [reply core/vm/actor/builtin/types/market/deal.hpp:62]: If you need a full copy, please write copy constructor and operator for `ClientDealProposal` where call `Universal::copy`

File: core/api/rpc/json.hpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 #include "primitives/cid/cid_of_cbor.hpp"
 #include "primitives/json_types.hpp"
 #include "vm/actor/builtin/types/market/deal.hpp"
+#include "vm/actor/builtin/types/market/v0/deal_proposal.hpp"
```
Comment (wer1st) [inline core/api/rpc/json.hpp:29]:
```suggestion
#include "vm/actor/builtin/types/market/deal_proposal.hpp"
```

File: core/api/rpc/json.hpp [hunk 5]
```diff
@@ -976,6 +977,42 @@ namespace fc::vm::actor::builtin::types {
...
+      Get(j, "EndEpoch", deal_proposal.end_epoch);
+      Get(j, "StoragePricePerEpoch", deal_proposal.storage_price_per_epoch);
+      Get(j, "ProviderCollateral", deal_proposal.provider_collateral);
+      Get(j, "ClientCollateral", deal_proposal.client_collateral);
+
+      v = vm::actor::builtin::types::Universal<
+          vm::actor::builtin::types::market::DealProposal>(
+          vm::actor::ActorVersion::kVersion0,
+          std::make_shared<vm::actor::builtin::v0::market::DealProposal>(
+              deal_proposal));
```
Comment (wer1st) [inline core/api/rpc/json.hpp:1013]:
```suggestion
      Universal<DealProposal> v{ActorVersion::kVersion0};
      
      Get(j, "PieceCID", v->piece_cid);
      Get(j, "PieceSize", v->piece_size);
      Get(j, "VerifiedDeal", v->verified);
      Get(j, "Client", v->client);
      Get(j, "Provider", v->provider);
      Get(j, "StartEpoch", v->start_epoch);
      Get(j, "EndEpoch", v->end_epoch);
      Get(j, "StoragePricePerEpoch", v->storage_price_per_epoch);
      Get(j, "ProviderCollateral", v->provider_collateral);
      Get(j, "ClientCollateral", v->client_collateral);
```

File: core/vm/actor/builtin/types/universal/universal.hpp (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/types/universal/universal.hpp]: There was an issue with cbor encoding, but it was fixed in cbor.

File: core/vm/actor/builtin/types/market/deal_proposal.hpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  using primitives::TokenAmount;
+  using primitives::address::Address;
+  using primitives::piece::PaddedPieceSize;
+
+  /**
+   * According FIP-0027 label is a string or bytes from actors v8.
+   */
+  class Label : public boost::variant<std::string, Bytes> {
+   public:
+    Label() : boost::variant<std::string, Bytes>("") {}
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/types/market/deal_proposal.hpp:30]: Default constructor is required for decoding, the value of default constructor is valid, while construction from non-utf8 string as string label must not be enabled and this is checked in `make` methods.

[the remark this answers is not in the document]
Comment (wer1st) [reply core/vm/actor/builtin/types/market/deal_proposal.hpp:30]: Ok

File: core/api/rpc/json.hpp [hunk 7]
```diff
@@ -976,6 +977,34 @@ namespace fc::vm::actor::builtin::types {
       Get(j, "SlashEpoch", v.slash_epoch);
     }
 
+    JSON_ENCODE(Universal<DealProposal>) {
+      Value j{rapidjson::kObjectType};
+      Set(j, "PieceCID", v->piece_cid, allocator);
```
Comment (wer1st) [inline core/api/rpc/json.hpp:982]:
```suggestion
      Value j{rapidjson::kObjectType};
      v = Universal<DealProposal>{ActorVersion::kVersion0};
      Set(j, "PieceCID", v->piece_cid, allocator);
```
